File: src/lib/uploadProviders.ts
export type UploadProviderKind = 'catbox' | 'filegarden' | 'http_multipart';

export type CatboxConfig = {
  apiUrl: string;
  userHash: string | null;
};

export type FilegardenConfig = {
  apiBase: string;
  email: string;
  password: string;
  totp: string | null;
  sessionCookie: string | null;
  authUserId: string | null;
  authToken: string | null;
  uploadUrl: string | null;
};

export type HttpMultipartConfig = {
  url: string;
  fileField: string;
  extraFields: Record<string, string>;
  headers: Record<string, string>;
  responseMode: 'plain_url' | 'json_path';
  responseJsonPath: string | null;
};

export type UploadProvider = {
  id: string;
  name: string;
  enabled: boolean;
  kind: UploadProviderKind;
  config: CatboxConfig | FilegardenConfig | HttpMultipartConfig;
};

export type UploadProviderSummary = {
  id: string;
  name: string;
  enabled: boolean;
  kind: UploadProviderKind;
  isDefault: boolean;
};

export type UploadProvidersEditorState = {
  providers: UploadProvider[];
  defaultUploadProviderId: string | null;
};

export function kindLabel(kind: UploadProviderKind): string {
  switch (kind) {
    case 'catbox':
      return 'Catbox';
    case 'filegarden':
      return 'File Garden';
    case 'http_multipart':
      return 'Custom HTTP';
    default:
      return kind;
  }
}

export function newProviderId(): string {
  return `upload-${Date.now().toString(36)}`;
}

export function createCatboxProvider(): UploadProvider {
  return {
    id: 'catbox',
    name: 'Catbox',
    enabled: true,
    kind: 'catbox',
    config: {
      apiUrl: 'https://catbox.moe/user/api.php',
      userHash: null,
    },
  };
}

export function createFilegardenProvider(): UploadProvider {
  return {
    id: 'filegarden',
    name: 'File Garden',
    enabled: true,
    kind: 'filegarden',
    config: {
      apiBase: 'https://api.filegarden.com',
      email: '',
      password: '',
      totp: null,
      sessionCookie: null,
      authUserId: null,
      authToken: null,
      uploadUrl: null,
    },
  };
}

export function createCustomProvider(): UploadProvider {
  return {
    id: newProviderId(),
    name: 'Custom server',
    enabled: true,
    kind: 'http_multipart',
    config: {
      url: '',
      fileField: 'file',
      extraFields: {},
      headers: {},
      responseMode: 'plain_url',
      responseJsonPath: null,
    },
  };
}

export function readUploadProvidersFromAppSettings(settings: Record<string, unknown>): UploadProvider[] {
  const raw = settings.uploadProviders ?? settings.upload_providers;
  return parseProvidersFromSettings(raw);
}

export function readDefaultUploadProviderId(settings: Record<string, unknown>): string | null {
  const raw = settings.defaultUploadProviderId ?? settings.default_upload_provider_id;
  return typeof raw === 'string' && raw.trim() ? raw.trim() : null;
}

export function normalizeUploadSummary(entry: unknown): UploadProviderSummary | null {
  if (!entry || typeof entry !== 'object') {
    return null;
  }

  const record = entry as Record<string, unknown>;
  const id = typeof record.id === 'string' ? record.id.trim() : '';
  const rawKind = typeof record.kind === 'string' ? record.kind.trim().toLowerCase() : '';

  if (!id || (rawKind !== 'catbox' && rawKind !== 'filegarden' && rawKind !== 'http_multipart')) {
    return null;
  }

  return {
    id,
    name: typeof record.name === 'string' && record.name.trim() ? record.name.trim() : id,
    enabled: record.enabled !== false,
    kind: rawKind as UploadProviderKind,
    isDefault: Boolean(record.isDefault ?? record.is_default),
  };
}

export function normalizeUploadSummaries(raw: unknown): UploadProviderSummary[] {
  if (!Array.isArray(raw)) {
    return [];
  }

  return raw
    .map((entry) => normalizeUploadSummary(entry))
    .filter((entry): entry is UploadProviderSummary => entry !== null);
}

export function parseProvidersFromSettings(raw: unknown): UploadProvider[] {
  if (!Array.isArray(raw)) {
    return [];
  }

  return raw
    .map((entry) => normalizeProvider(entry))
    .filter((entry): entry is UploadProvider => entry !== null);
}

function normalizeProvider(entry: unknown): UploadProvider | null {
  if (!entry || typeof entry !== 'object') {
    return null;
  }

  const record = entry as Record<string, unknown>;
  const id = typeof record.id === 'string' ? record.id.trim() : '';
  const name = typeof record.name === 'string' ? record.name.trim() : '';
  const rawKind = typeof record.kind === 'string' ? record.kind.trim().toLowerCase() : '';
  const config = record.config;

  if (!id || (rawKind !== 'catbox' && rawKind !== 'filegarden' && rawKind !== 'http_multipart')) {
    return null;
  }

  const kind = rawKind as UploadProviderKind;

  return {
    id,
    name: name || id,
    enabled: record.enabled !== false,
    kind,
    config: normalizeConfig(kind, config),
  };
}

function normalizeConfig(
  kind: UploadProviderKind,
  config: unknown,
): CatboxConfig | FilegardenConfig | HttpMultipartConfig {
  const record = config && typeof config === 'object' ? (config as Record<string, unknown>) : {};

  if (kind === 'catbox') {
    return {
      apiUrl:
        typeof record.apiUrl === 'string' && record.apiUrl.trim()
          ? record.apiUrl.trim()
          : 'https://catbox.moe/user/api.php',
      userHash: typeof record.userHash === 'string' && record.userHash.trim() ? record.userHash.trim() : null,
    };
  }

  if (kind === 'filegarden') {
    const rawApiBase =
      typeof record.apiBase === 'string' && record.apiBase.trim()
        ? record.apiBase.trim()
        : 'https://api.filegarden.com';
    const apiBase =
      rawApiBase === 'https://filegarden.com' ||
      rawApiBase === 'http://filegarden.com' ||
      rawApiBase.startsWith('https://www.filegarden.com') ||
      rawApiBase.startsWith('http://www.filegarden.com')
        ? 'https://api.filegarden.com'
        : rawApiBase;
    const uploadUrl =
      typeof record.uploadUrl === 'string' && record.uploadUrl.trim()
        ? record.uploadUrl.trim()
        : null;
    return {
      apiBase,
      email: typeof record.email === 'string' ? record.email : '',
      password: typeof record.password === 'string' ? record.password : '',
      totp: typeof record.totp === 'string' && record.totp.trim() ? record.totp.trim() : null,
      sessionCookie:
        typeof record.sessionCookie === 'string' && record.sessionCookie.trim()
          ? record.sessionCookie.trim()
          : null,
      authUserId:
        typeof record.authUserId === 'string' && record.authUserId.trim()
          ? record.authUserId.trim()
          : null,
      authToken:
        typeof record.authToken === 'string' && record.authToken.trim() ? record.authToken.trim() : null,
      uploadUrl:
        uploadUrl && uploadUrl.includes('/api/v0/files') ? null : uploadUrl,
    };
  }

  return {
    url: typeof record.url === 'string' ? record.url : '',
    fileField: typeof record.fileField === 'string' && record.fileField.trim() ? record.fileField.trim() : 'file',
    extraFields: stringMap(record.extraFields),
    headers: stringMap(record.headers),
    responseMode: record.responseMode === 'json_path' ? 'json_path' : 'plain_url',
    responseJsonPath:
      typeof record.responseJsonPath === 'string' && record.responseJsonPath.trim()
        ? record.responseJsonPath.trim()
        : null,
  };
}

function stringMap(value: unknown): Record<string, string> {
  if (!value || typeof value !== 'object') {
    return {};
  }

  const result: Record<string, string> = {};
  for (const [key, entry] of Object.entries(value as Record<string, unknown>)) {
    if (typeof entry === 'string') {
      result[key] = entry;
    }
  }
  return result;
}

export function serializeProviders(providers: UploadProvider[]): UploadProvider[] {
  return providers.map((provider) => ({
    ...provider,
    config: { ...provider.config },
  }));
}

File: src-tauri/src/upload_providers.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;

pub const KIND_CATBOX: &str = "catbox";
pub const KIND_FILEGARDEN: &str = "filegarden";
pub const KIND_HTTP_MULTIPART: &str = "http_multipart";

pub const DEFAULT_CATBOX_ID: &str = "catbox";
pub const DEFAULT_FILEGARDEN_ID: &str = "filegarden";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadProvider {
    pub id: String,
    pub name: String,
    pub enabled: bool,
    pub kind: String,
    pub config: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UploadProviderSummary {
    pub id: String,
    pub name: String,
    pub enabled: bool,
    pub kind: String,
    pub is_default: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CatboxConfig {
    #[serde(default = "default_catbox_api_url")]
    pub api_url: String,
    pub user_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilegardenConfig {
    #[serde(default = "default_filegarden_api_base")]
    pub api_base: String,
    pub email: String,
    pub password: String,
    pub totp: Option<String>,
    #[serde(default)]
    pub session_cookie: Option<String>,
    #[serde(default)]
    pub auth_user_id: Option<String>,
    #[serde(default)]
    pub auth_token: Option<String>,
    #[serde(default)]
    pub upload_url: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HttpMultipartConfig {
    pub url: String,
    #[serde(default = "default_file_field")]
    pub file_field: String,
    #[serde(default)]
    pub extra_fields: HashMap<String, String>,
    #[serde(default)]
    pub headers: HashMap<String, String>,
    #[serde(default = "default_response_mode")]
    pub response_mode: String,
    pub response_json_path: Option<String>,
}

fn default_catbox_api_url() -> String {
    "https://catbox.moe/user/api.php".to_string()
}

fn default_filegarden_api_base() -> String {
    "https://api.filegarden.com".to_string()
}

fn default_file_field() -> String {
    "file".to_string()
}

fn default_response_mode() -> String {
    "plain_url".to_string()
}

pub fn default_catbox_provider() -> UploadProvider {
    UploadProvider {
        id: DEFAULT_CATBOX_ID.to_string(),
        name: "Catbox".to_string(),
        enabled: true,
        kind: KIND_CATBOX.to_string(),
        config: json!({
            "apiUrl": default_catbox_api_url(),
            "userHash": null
        }),
    }
}

pub fn default_filegarden_provider() -> UploadProvider {
    UploadProvider {
        id: DEFAULT_FILEGARDEN_ID.to_string(),
        name: "File Garden".to_string(),
        enabled: true,
        kind: KIND_FILEGARDEN.to_string(),
        config: json!({
            "apiBase": default_filegarden_api_base(),
            "email": "",
            "password": "",
            "totp": null,
            "sessionCookie": null,
            "authUserId": null,
            "authToken": null,
            "uploadUrl": null
        }),
    }
}

pub fn migrate_upload_providers(
    providers: &mut Vec<UploadProvider>,
    default_id: &mut Option<String>,
    catbox_user_hash: &Option<String>,
    catbox_api_url: &Option<String>,
) {
    if !providers.is_empty() {
        return;
    }

    let has_legacy = catbox_user_hash
        .as_ref()
        .is_some_and(|v| !v.trim().is_empty())
        || catbox_api_url
            .as_ref()
            .is_some_and(|v| !v.trim().is_empty());

    if has_legacy {
        let mut provider = default_catbox_provider();
        if let Some(url) = catbox_api_url.as_ref().filter(|v| !v.trim().is_empty()) {
            provider.config["apiUrl"] = json!(url.trim());
        }
        if let Some(hash) = catbox_user_hash.as_ref().filter(|v| !v.trim().is_empty()) {
            provider.config["userHash"] = json!(hash.trim());
        }
        providers.push(provider);
    } else {
        providers.push(default_filegarden_provider());
    }

    if default_id.is_none() {
        *default_id = providers.first().map(|provider| provider.id.clone());
    }
}

pub fn normalize_settings_providers(
    providers: &mut [UploadProvider],
    default_id: &mut Option<String>,
) -> Result<(), String> {
    if providers.is_empty() {
        return Err("Add at least one upload target in Settings.".to_string());
    }

    let mut seen = std::collections::HashSet::new();
    for provider in providers.iter() {
        let id = provider.id.trim();
        if id.is_empty() {
            return Err("Each upload provider needs a non-empty id.".to_string());
        }
        if !seen.insert(id.to_string()) {
            return Err(format!("Duplicate upload provider id: {id}"));
        }
    }

    for provider in providers.iter_mut() {
        provider.id = provider.id.trim().to_string();
        provider.name = provider.name.trim().to_string();
        if provider.name.is_empty() {
            provider.name = provider.id.clone();
        }
        provider.kind = provider.kind.trim().to_lowercase();
    }

    let first_enabled_id = || {
        providers
            .iter()
            .find(|provider| provider.enabled)
            .map(|provider| provider.id.clone())
            .or_else(|| providers.first().map(|provider| provider.id.clone()))
    };

    if let Some(default) = default_id.clone() {
        let default_valid = providers
            .iter()
            .any(|provider| provider.id == default && provider.enabled);
        if !default_valid {
            *default_id = first_enabled_id();
        }
    } else {
        *default_id = first_enabled_id();
    }

    Ok(())
}

pub fn provider_summary(
    provider: &UploadProvider,
    default_id: &Option<String>,
) -> UploadProviderSummary {
    UploadProviderSummary {
        id: provider.id.clone(),
        name: provider.name.clone(),
        enabled: provider.enabled,
        kind: provider.kind.clone(),
        is_default: default_id.as_ref() == Some(&provider.id),
    }
}

pub fn parse_catbox_config(config: &Value) -> Result<CatboxConfig, String> {
    serde_json::from_value(config.clone())
        .map_err(|err| format!("Invalid Catbox provider config: {err}"))
}

pub fn parse_filegarden_config(config: &Value) -> Result<FilegardenConfig, String> {
    serde_json::from_value(config.clone())
        .map_err(|err| format!("Invalid File Garden provider config: {err}"))
}

pub fn parse_http_multipart_config(config: &Value) -> Result<HttpMultipartConfig, String> {
    serde_json::from_value(config.clone())
        .map_err(|err| format!("Invalid custom upload provider config: {err}"))
}

pub fn filegarden_config_to_value(config: &FilegardenConfig) -> Value {
    serde_json::to_value(config).unwrap_or_else(|_| json!({}))
}
